################################################################################
SHORT DESCRIPTION: 
################################################################################
XSA-139 (http://xenbits.xen.org/xsa/advisory-139.html)
Use after free in QEMU/Xen block unplug protocol.

################################################################################
LONG DESCRIPTION: 
################################################################################
Source: http://xenbits.xen.org/xsa/advisory-139.html
Patches: xsa139-qemuu-4.5.patch

pci_piix3_xen_ide_unplug should completely unhook the unplugged IDEDevice from
the corresponding BlockBackend, otherwise the next call to release_drive will
try to detach the drive again.

################################################################################
CHANGELOG 
################################################################################
Added in OpenXT, Xen 4.3.4 patch-queue: Eric Chanudet <chanudete@ainfosec.com>

################################################################################
PATCHES 
################################################################################
--- a/hw/ide/piix.c
+++ b/hw/ide/piix.c
@@ -172,6 +172,7 @@ static int pci_piix3_xen_ide_unplug(Devi
     PCIIDEState *pci_ide;
     DriveInfo *di;
     int i = 0;
+    IDEDevice *idedev;
 
     pci_dev = DO_UPCAST(PCIDevice, qdev, dev);
     pci_ide = DO_UPCAST(PCIIDEState, dev, pci_dev);
@@ -185,6 +186,12 @@ static int pci_piix3_xen_ide_unplug(Devi
             }
             bdrv_close(di->bdrv);
             pci_ide->bus[di->bus].ifs[di->unit].bs = NULL;
+            if (!(i % 2)) {
+                idedev = pci_ide->bus[di->bus].master;
+            } else {
+                idedev = pci_ide->bus[di->bus].slave;
+            }
+            idedev->conf.bs = NULL;
             drive_put_ref(di);
         }
     }
